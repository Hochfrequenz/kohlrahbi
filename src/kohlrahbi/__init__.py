"""
kohlrahbi is a package to scrape AHBs (in docx format)
"""
import gc
import re
import sys
from pathlib import Path
from typing import Any

import click
import docx  # type:ignore[import]
import tomlkit

from kohlrahbi.ahb.ahbtable import AhbTable
from kohlrahbi.ahbfilefinder import AhbFileFinder
from kohlrahbi.logger import logger
from kohlrahbi.read_functions import get_ahb_table
from kohlrahbi.unfoldedahb.unfoldedahbtable import UnfoldedAhb

_pruefi_pattern = re.compile(r"^[1-9]\d{4}$")


def get_valid_pruefis(list_of_pruefis: list[str]) -> list[str]:
    """
    This function returns a new list with only those pruefis which match the pruefi_pattern.
    """
    valid_pruefis: list[str] = [pruefi for pruefi in list_of_pruefis if _pruefi_pattern.match(pruefi)]
    return valid_pruefis


def check_python_version():
    """
    Check if the Python interpreter is greater or equal to 3.11
    """
    if sys.version_info.major != 3 or sys.version_info.minor < 11:
        raise click.Abort(
            f"""Python >=3.11 is required to run this script but you use Python
{sys.version_info.major}.{sys.version_info.minor}"""
        )


def check_output_path(path: Path) -> None:
    """
    Checks if the given path exists and if not it asks the user if they want to create the given directory.
    """
    if not path.exists():
        click.secho("‚ö†Ô∏è The output directory does not exist.", fg="red")

        if click.confirm(f"Should I try to create the directory at '{path}'?", default=True):
            try:
                path.mkdir(exist_ok=True)
                click.secho(f"üìÇ The output directory is created at {path.absolute()}.", fg="yellow")
            except FileNotFoundError as fnfe:
                click.secho(
                    "üò± There was an path error. I can only create a new directory in an already existing directory.",
                    fg="red",
                )
                click.secho(f"Your given path is '{path}'", fg="red")
                click.secho(str(fnfe), fg="red")
                raise click.Abort()

        else:
            click.secho("üëã Alright I will end this program now. Have a nice day.", fg="green")
            raise click.Abort()


def load_all_known_pruefis_from_file(
    path_to_all_known_pruefis: Path = Path(__file__).parent / Path("all_known_pruefis.toml"),
) -> list[str]:
    """
    Loads the file which contains all known Pr√ºfidentifikatoren.
    The file may be manually updated with the script `collect_pruefis.py`.
    """

    with open(path_to_all_known_pruefis, "rb") as file:
        state_of_kohlrahbi: dict[str, Any] = tomlkit.load(file)

    meta_data_section = state_of_kohlrahbi.get("meta_data")
    content_section = state_of_kohlrahbi.get("content")

    if meta_data_section is None:
        click.secho(f"There is no 'meta_data' section in the provided toml file: {path_to_all_known_pruefis}", fg="red")
        raise click.Abort()
    if content_section is None:
        click.secho(f"There is no 'content' section in the toml file: {path_to_all_known_pruefis}", fg="red")
        raise click.Abort()

    pruefis: list[str] = content_section.get("pruefidentifikatoren")
    return pruefis


@click.command()
@click.option(
    "-p",
    "--pruefis",
    default=[],
    required=False,
    help="Five digit number like 11042.",
    multiple=True,
)
@click.option(
    "-i",
    "--input_path",
    type=click.Path(exists=True, dir_okay=True, file_okay=False, path_type=Path),
    prompt="Input directory",
    help="Define the path to the folder with the docx AHBs.",
)
@click.option(
    "-o",
    "--output_path",
    type=click.Path(exists=False, dir_okay=True, file_okay=False, path_type=Path),
    default="output",
    prompt="Output directory",
    help="Define the path where you want to save the generated files.",
)
@click.option(
    "--file-type",
    type=click.Choice(["flatahb", "csv", "xlsx"], case_sensitive=False),
    multiple=True,
)
@click.option(
    "--assume-yes",
    "-y",
    is_flag=True,
    help="Confirm all prompts automatically.",
)
# pylint: disable=too-many-branches, too-many-statements, too-many-locals
def main(pruefis: list[str], input_path: Path, output_path: Path, file_type: list[str], assume_yes: bool):
    """
    A program to get a machine readable version of the AHBs docx files published by edi@energy.
    """
    check_python_version()

    if not assume_yes:
        check_output_path(path=output_path)
    else:
        if output_path.exists():
            click.secho(f"The output directory '{output_path}' exists already.", fg="yellow")
        else:
            output_path.mkdir(parents=True)
            click.secho(f"I created a new directory at {output_path}", fg="yellow")

    if not any(pruefis):
        click.secho("‚òùÔ∏è No pruefis were given. I will parse all known pruefis.", fg="yellow")
        pruefis = load_all_known_pruefis_from_file()
    if not any(file_type):
        message = "‚Ñπ You did not provide any value for the parameter --file-type. No files will be created."
        click.secho(message, fg="yellow")
        logger.warning(message)

    valid_pruefis: list[str] = get_valid_pruefis(list_of_pruefis=pruefis)
    if not any(valid_pruefis):
        click.secho("‚ö†Ô∏è There are no valid pruefidentifkatoren.", fg="red")
        raise click.Abort()

    if len(valid_pruefis) != len(pruefis):
        click.secho("‚òùÔ∏è Not all given pruefidentifikatoren are valid.", fg="yellow")
        click.secho(f"I will continue with the following valid pruefis: {valid_pruefis}.", fg="yellow")
    path_to_document_mapping: dict[Path, docx.Document] = {}

    for pruefi in valid_pruefis:
        try:
            logger.info("start looking for pruefi '%s'", pruefi)
            ahb_file_finder = AhbFileFinder.from_input_path(input_path=input_path)
            ahb_file_paths: list[Path] = ahb_file_finder.get_docx_files_which_may_contain_searched_pruefi(
                searched_pruefi=pruefi
            )

            if not any(ahb_file_paths):
                logger.warning("No docx file was found for pruefi '%s'", pruefi)
                continue

            for ahb_file_path in ahb_file_paths:
                if not (doc := path_to_document_mapping.get(ahb_file_path, None)):
                    try:
                        doc = docx.Document(ahb_file_path)  # Creating word reader object.
                        path_to_document_mapping[ahb_file_path] = doc
                        logger.debug("Saved %s document in cache", ahb_file_path)  # to not re-read it every time
                    except IOError as ioe:
                        logger.exception("There was an error opening the file '%s'", ahb_file_path, exc_info=True)
                        raise click.Abort() from ioe

                logger.info("start reading docx file '%s'", str(ahb_file_path))

                ahb_table: AhbTable | None = get_ahb_table(
                    document=doc,
                    pruefi=pruefi,
                )

                if ahb_table is None:
                    continue

                if isinstance(ahb_table, AhbTable):
                    unfolded_ahb = UnfoldedAhb.from_ahb_table(ahb_table=ahb_table, pruefi=pruefi)

                    if "xlsx" in file_type:
                        logger.info("üíæ Saving xlsx file %s", pruefi)
                        unfolded_ahb.dump_xlsx(path_to_output_directory=output_path)

                    if "flatahb" in file_type:
                        logger.info("üíæ Saving flatahb file %s", pruefi)
                        unfolded_ahb.dump_flatahb_json(output_directory_path=output_path)

                    if "csv" in file_type:
                        logger.info("üíæ Saving csv file %s", pruefi)
                        unfolded_ahb.dump_csv(path_to_output_directory=output_path)
                    break
        except Exception as general_error:  # pylint:disable=broad-except
            logger.exception(
                "There was an uncaught error while processing the pruefi '%s': %s",
                pruefi,
                str(general_error),
                exc_info=True,
            )
            continue
        del ahb_table
        if "unfolded_ahb" in locals():
            del unfolded_ahb
        gc.collect()


if __name__ == "__main__":
    # the parameter arguments gets provided over the CLI
    main()  # pylint:disable=no-value-for-parameter
