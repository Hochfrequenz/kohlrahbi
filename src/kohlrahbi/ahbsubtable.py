import attrs
import pandas as pd
from docx.table import Table as DocxTable  # type:ignore[import]
from docx.table import _Cell  # type:ignore[import]

from kohlrahbi.ahbtablerow import AhbTableRow
from kohlrahbi.row_type_checker import RowType, get_row_type
from kohlrahbi.seed import Seed


@attrs.define(auto_attribs=True, kw_only=True)
class AhbSubTable:
    """
    The AHB table for one Pruefidentifikator is separated into small sub tables.
    This class contains the information from such a sub table.
    """

    table_meta_data: Seed
    table: pd.DataFrame

    @classmethod
    def from_table(cls, docx_table: DocxTable) -> "AhbSubTable":
        """
        Create a new AhbSubTable instance from a docx table
        """

        ahb_table_meta_data = Seed.from_table(docx_table=docx_table)

        # ahb_sub_table = cls.parse(docx_table=docx_table, ahb_table_meta_data=table_meta_data)

        ahb_table_dataframe = pd.DataFrame(
            columns=ahb_table_meta_data.column_headers,
            dtype="str",
        )

        for row in docx_table.rows:
            sanitized_row = list(AhbSubTable._iter_visible_cells(row=row))

            current_edifact_struktur_cell = sanitized_row[0]

            # check for row type
            current_row_type = get_row_type(
                edifact_struktur_cell=current_edifact_struktur_cell,
                left_indent_position=ahb_table_meta_data.edifact_struktur_left_indent_position,
            )

            edifact_struktur_cell = sanitized_row[0]
            middle_cell = sanitized_row[1]
            bedingung_cell = sanitized_row[-1]

            # this case covers the "normal" docx table row
            if not (current_row_type is RowType.EMPTY and ahb_table_meta_data.last_two_row_types[0] is RowType.HEADER):
                ahb_table_row = AhbTableRow(
                    seed=ahb_table_meta_data,
                    edifact_struktur_cell=edifact_struktur_cell,
                    middle_cell=middle_cell,
                    bedingung_cell=bedingung_cell,
                )

                ahb_table_row_dataframe = ahb_table_row.parse(row_type=current_row_type)

                if ahb_table_row_dataframe is not None:
                    ahb_table_dataframe = pd.concat([ahb_table_dataframe, ahb_table_row_dataframe], ignore_index=True)
            # this case covers the page break situation
            else:
                ahb_table_row = AhbTableRow(
                    seed=ahb_table_meta_data,
                    edifact_struktur_cell=edifact_struktur_cell,
                    middle_cell=middle_cell,
                    bedingung_cell=bedingung_cell,
                )

                ahb_table_row.parse(row_type=ahb_table_meta_data.last_two_row_types[1])

            ahb_table_meta_data.last_two_row_types[1] = ahb_table_meta_data.last_two_row_types[0]
            ahb_table_meta_data.last_two_row_types[0] = current_row_type

        return cls(table_meta_data=ahb_table_meta_data, table=ahb_table_dataframe)

    @staticmethod
    def _iter_visible_cells(row):
        """
        This function makes sure that you will iterate over the cells you see in the word document.
        For more information go to https://github.com/python-openxml/python-docx/issues/970#issuecomment-877386927
        """
        table_row = row._tr  # pylint:disable=protected-access
        for table_column in table_row.tc_lst:
            yield _Cell(table_column, row.table)
