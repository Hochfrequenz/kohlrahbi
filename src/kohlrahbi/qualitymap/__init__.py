"""
This module contains the functions to scrape the AHBs for quality maps.
"""

from pathlib import Path
from typing import Optional

import docx
import pandas as pd
from docx.document import Document
from docx.table import Table

from kohlrahbi.docxfilefinder import DocxFileFinder
from kohlrahbi.logger import logger
from kohlrahbi.qualitymap.qualitymaptable import QualityMapTable
from kohlrahbi.read_functions import get_all_paragraphs_and_tables


def find_docx_files(input_path: Path) -> list[Path]:
    """
    Find all .docx files containing quality maps.
    """
    docx_file_finder = DocxFileFinder.from_input_path(input_path=input_path)
    return docx_file_finder.get_docx_files_which_contain_quality_map()


def is_quality_map_table(table: Table) -> bool:
    """
    Checks if the given table is quality map table.
    """
    try:
        return table.cell(row_idx=0, col_idx=0).text.strip() == "Qualität\n\nSegmentgruppe"
    except IndexError:
        return False


def get_quality_map_table(document: Document) -> Optional[QualityMapTable]:
    """
    Reads a docx file and extracts the quality map table.
    Returns None if no such table was found.
    """

    # Iterate through the whole word document
    logger.info("🔁 Start iterating through paragraphs and tables")
    for table in document.tables:
        if is_quality_map_table(table=table):
            change_history_table = QualityMapTable.from_docx_quality_map_table(docx_table=table)
            return change_history_table

    return None


def process_docx_file(file_path: Path) -> Optional[QualityMapTable]:
    """
    Read and process quality map from a .docx file.
    """
    doc = docx.Document(str(file_path))
    logger.info("🤓 Start reading docx file '%s'", str(file_path))
    quality_map_table = get_quality_map_table(document=doc)

    if quality_map_table is not None:
        return quality_map_table
    return None


def scrape_quality_map(input_path: Path, output_path: Path) -> None:
    """
    starts the scraping process of the quality maps.

    Args:
        input_path (Path): The path to the input **directory** containing the docx files with quality maps.
        output_path (Path): The path to the output **directory** to save the scraped quality maps
    """
    if input_path.is_dir() is False:
        logger.error("The input path '%s' is not a directory.", input_path)
        raise NotADirectoryError(f"The input path '{input_path}' is not a directory.")

    logger.info("👀 Start looking for quality maps")
    ahb_file_paths = find_docx_files(input_path)

    for file_path in ahb_file_paths:
        quality_map_table = process_docx_file(file_path)
        logger.info("Done with %s", file_path)
        if quality_map_table is None:
            logger.info("No quality map table found in %s", file_path)
            continue
        quality_map_table.save_to_csv(output_path / Path(f"{file_path.stem}_quality_map.csv"))
        quality_map_table.save_to_xlsx(output_path / Path(f"{file_path.stem}_quality_map.xlsx"))
