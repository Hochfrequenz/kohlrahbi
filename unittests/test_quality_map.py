import glob
from pathlib import Path
from typing import Union
from unittest.mock import MagicMock

import pytest
from click.testing import CliRunner, Result

from kohlrahbi import cli
from kohlrahbi.qualitymap import is_quality_map_table

runner: CliRunner = CliRunner()


class MockCell:
    def __init__(self, text):
        self.text = text

    def strip(self):
        return self.text.strip()


class MockTable:
    def __init__(self, cell_text):
        self.cell_text = cell_text

    def cell(self, row_idx, col_idx):
        if row_idx == 0 and col_idx == 0:
            return MockCell(self.cell_text)
        raise IndexError


@pytest.mark.snapshot
class TestQualityMap:
    """
    This class contains the unit tests for the quality map
    """

    def test_is_quality_map_table_true(self):
        table = MockTable("QualitÃ¤t\n\nSegmentgruppe")
        assert is_quality_map_table(table) is True

    def test_is_quality_map_table_false(self):
        table = MockTable("Some other text")
        assert is_quality_map_table(table) is False

    def test_is_quality_map_table_index_error(self):
        table = MockTable(None)
        table.cell = MagicMock(side_effect=IndexError)
        assert is_quality_map_table(table) is False

    @pytest.mark.parametrize(
        "argument_options, expected_response",
        [
            pytest.param(
                [
                    "qualitymap",
                    "--format-version",
                    "FV2504",
                ],
                {"exit_code": 0, "output_snippet": ""},
                id="check command for quality map table",
            )
        ],
    )
    def test_cli_quality_map_table(
        self, argument_options: list[str], expected_response: dict[str, Union[str, int]], snapshot, tmp_path
    ):

        actual_output_dir = tmp_path / "actual-output"

        argument_options.extend(
            [
                "--assume-yes",
                "--edi-energy-mirror-path",
                str(Path(__file__).parents[1] / "edi_energy_mirror"),
                "--output-path",
                str(actual_output_dir),
            ]
        )

        # Call the CLI tool with the desired arguments
        response: Result = runner.invoke(cli, argument_options)

        assert response.exit_code == expected_response.get("exit_code")

        # Check if the generated files are the same as the expected files

        def get_csv_file(directory: Path) -> Path:
            csv_files = glob.glob(str(directory / "*.csv"))
            if csv_files:
                return Path(csv_files[0])
            else:
                raise FileNotFoundError("No CSV file found in the given directory")

        path_to_actual_csv_file = get_csv_file(actual_output_dir)

        with open(path_to_actual_csv_file, "r", encoding="utf-8") as actual_csv:
            assert snapshot == actual_csv.read()
