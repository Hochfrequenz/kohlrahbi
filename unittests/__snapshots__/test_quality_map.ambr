# serializer version: 1
# name: TestQualityMap.test_cli_quality_map_table[check command for quality map table]
  '''
  Qualität \ Segmentgruppe,Bestellte Daten,Gültige Daten,Informative Daten,Erwartete Daten,Im System vorhandene Daten
  "SG6 RFF DE1153
  Termine der Marktlokation",--,"Z50
  „Termindaten der Marktlokation“",--,"Z51
  „Erwartete Termindaten der Marktlokation“","Z52
  „Im System vorhandene Termindaten der Marktlokation“"
  "SG8 SEQ DE1229
  Referenz auf die Lokationsbündelstruktur",--,"Z78
  „Referenz auf die Lokationsbündelstruktur“","ZD5
  „Informative Referenz auf die Lokationsbündelstruktur“","ZC7
  „Erwartete Referenz auf die Lokationsbündelstruktur“","ZC8
  „Im System vorhandene Referenz auf die Lokationsbündelstruktur“"
  "SG8 SEQ DE1229
  Objektcode der Lokation in der Lokationsbündelstruktur",--,"Z58
  „Zuordnung Lokation zum Objektcode des Lokationsbündels“","ZD6
  „Informative Zuordnung Lokation zum Objektcode des Lokationsbündels“","ZC9
  „Erwartete Zuordnung Lokation zum Objektcode des Lokationsbündels“","ZD0
  „Im System vorhandene Zuordnung Lokation zum Objektcode des Lokationsbündels“"
  "SG8 SEQ DE1229
  Bestandteil eines Produktpakets","Z79
  „Bestandteil eines Produktpakets“",--,--,--,--
  "SG8 SEQ DE1229
  Priorisierung erforderliches Produktpaket","ZH0
  „Priorisierung erforderliches Produktpaket“",--,--,--,--
  "SG8 SEQ DE1229
  Daten der Netzlokation",--,"Z51
  „Daten der Netzlokation“","ZD7
  „Informative Daten der Netzlokation“","ZA9
  „Erwartete Daten der Netzlokation“","ZB0
  „Im System vorhandene Daten der Netzlokation“"
  "SG8 SEQ DE1229
  Abrechnungsdaten der Netzlokation",--,"Z71
  „Abrechnungsdaten der Netzlokation“","ZD8
  „Informative Abrechnungsdaten der Netzlokation“","ZH1
  „Erwartete Abrechnungsdaten der Netzlokation“","ZH2
  „Im System vorhandene Abrechnungsdaten der Netzlokation“"
  "SG8 SEQ DE1229
  OBIS-Daten der Netzlokation",--,"Z57
  „OBIS-Daten der Netzlokation“","ZD9
  „Informative OBIS-Daten der Netzlokation“","ZA7
  „Erwartete OBIS-Daten der Netzlokation“","ZA8
  „Im System vorhandene OBIS-Daten der Netzlokation“"
  "SG8 SEQ DE1229
  Produkt-Daten der Netzlokation",--,"Z60
  „Produkt-Daten der Netzlokation“","ZE0
  „Informative Produkt-Daten der Netzlokation“","ZG8
  „Erwartete Produkt-Daten der NeLo“","ZG9
  „Im System vorhandene Produkt-Daten der NeLo“"
  "SG8 SEQ DE1229
  Daten der Marktlokation",--,"Z01
  „Daten der Marktlokation“","Z98
  „Informative Daten der Marktlokation“","Z80
  „Erwartete Daten der Marktlokation“","Z81
  „Im System vorhandene Daten der Marktlokation“"
  "SG8 SEQ DE1229
  Daten der Marktlokation der beteiligten Marktrolle",--,--,--,--,"Z29
  „Daten der Marktlokation der beteiligten Marktrolle“"
  "SG8 SEQ DE1229
  Netznutzungsabrechnungsdaten der Marktlokation",--,"Z45
  „Netznutzungsabrechnungsdaten der Marktlokation“
  
  Z84
  „Differenz-Netznutzungsabrechnungsdaten der Marktlokation“","ZE1
  „Informative Netznutzungsabrechnungsdaten der Marktlokation“","Z82
  „Erwartete Netznutzungsabrechnungsdaten der Marktlokation“
  
  Z96
  „Erwartete Differenz-Netznutzungsabrechnungsdaten der Marktlokation“","Z83
  „Im System vorhandene Netznutzungsabrechnungsdaten der Marktlokation“
  
  Z97
  „Im System vorhandene Differenz-Netznutzungsabrechnungsdaten der Marktlokation“"
  "SG8 SEQ DE1229
  Messstellenbetriebsabrechnungsdaten der Marktlokation",--,"Z76
  „Messstellenbetriebsabrechnungsdaten der Marktlokation“",--,"ZC5
  „Erwartete Messstellenbetriebsabrechnungsdaten der Marktlokation“","ZC6
  „Im System vorhandene Messstellenbetriebsabrechnungsdaten der Marktlokation“"
  "SG8 SEQ DE1229
  Erforderliches Messprodukt der Marktlokation","Z27
  „Erforderliches Messprodukt der Marktlokation“",--,"ZE2
  „Informative Erforderliches Messprodukt der Marktlokation“",--,--
  "SG8 SEQ DE1229
  OBIS-Daten der Marktlokation",--,"Z02
  „OBIS-Daten der Marktlokation“","ZE3
  „Informative OBIS-Daten der Marktlokation“","ZA1
  „Erwartete OBIS-Daten der Marktlokation“","ZA2
  „Im System vorhandene OBIS-Daten der Marktlokation“"
  "SG8 SEQ DE1229
  Produkt-Daten der Marktlokation",--,"Z59
  „Produkt-Daten der Marktlokation“","ZE4
  „Informative Produkt-Daten der Marktlokation“","ZB5
  „Erwartete Produkt-Daten der Marktlokation“","ZB6
  „Im System vorhandene Produkt-Daten der Marktlokation“"
  "SG8 SEQ DE1229
  Verbrauchsart und Nutzung der OBIS-Kennzahl an der Marktlokation",--,"Z44
  „Verbrauchsart und Nutzung der OBIS-Kennzahl an der Marktlokation“","ZE5
  „Informative Verbrauchsart und Nutzung der OBIS-Kennzahl an der Marktlokation“","ZD1
  „Erwartete Verbrauchsart und Nutzung der OBIS-Kennzahl der Marktlokation“","ZD2
  „Im System vorhandene Verbrauchsart und Nutzung der OBIS-Kennzahl der Marktlokation“"
  "SG8 SEQ DE1229
  OBIS-Daten der Marktlokation der beteiligten Marktrolle",--,--,--,--,"Z30
  „OBIS-Daten der Marktlokation der beteiligten Marktrolle“"
  "SG8 SEQ DE1229
  Produkt-Daten der Marktlokation des NB",--,"Z40
  „Produkt-Daten der Marktlokation des NB“","ZE6
  „Informative Produkt-Daten der Marktlokation des NB“","ZD3
  „Erwartete Produkt-Daten der Marktlokation des NB“","ZD4
  „Im System vorhandene Produkt-Daten der Marktlokation des NB“"
  "SG8 SEQ DE1229
  Daten der Tranche",--,"Z15
  „Daten der Tranche“","ZE7
  „Informative Daten der Tranche“","Z94
  „Erwartete Daten der Tranche“","Z95
  „Im System vorhandene Daten der Tranche“"
  "SG8 SEQ DE1229
  Daten der Tranche der beteiligten Marktrolle",--,--,--,--,"Z31
  „Daten der Tranche der beteiligten Marktrolle“"
  "SG8 SEQ DE1229
  Erforderliches Messprodukt der Tranche","Z16
  „Erforderliches Produkt der Tranche“",--,"ZE8
  „Informative Erforderliches Produkt der Tranche“",--,--
  "SG8 SEQ DE1229
  OBIS-Daten der Tranche",--,"Z17
  „OBIS-Daten der Tranche“","ZE9
  „Informative OBIS-Daten der Tranche“","Z99
  „Erwartete OBIS-Daten der Tranche“","ZA0
  „Im System vorhandene OBIS-Daten der Tranche“"
  "SG8 SEQ DE1229
  OBIS-Daten der Tranche der beteiligten Marktrolle",--,--,--,--,"Z32
  „OBIS-Daten der Tranche der beteiligten Marktrolle“"
  "SG8 SEQ DE1229
  Daten der Technischen Ressource",--,"Z52
  „Daten der Technischen Ressource“","ZF0
  „Informative Daten der Technischen Ressource“","ZG4
  „Erwartete Daten der Technischen Ressource“","ZG5
  „Im System vorhandene Daten der Technischen Ressource“"
  "SG8 SEQ DE1229
  Daten der Steuerbaren Ressource",--,"Z62
  „Daten der Steuerbaren Ressource“","ZF1
  „Informative Daten der Steuerbaren Ressource“","ZB1
  „Erwartete Daten der Steuerbaren Ressource“","ZB2
  „Im System vorhandene Daten der Steuerbaren Ressource“"
  "SG8 SEQ DE1229
  Produkt-Daten der Steuerbaren Ressource",--,"Z61
  „Produkt-Daten der Steuerbaren Ressource“","ZF2
  „Informative Produkt-Daten der Steuerbaren Ressource“","ZB3
  „Erwartete Produkt-Daten der Steuerbaren Ressource“","ZB4
  „Im System vorhandene Produkt-Daten der Steuerbaren Ressource“"
  "SG8 SEQ DE1229
  Daten der Messlokation",--,"Z18
  „Daten der Messlokation“","ZF3
  „Informative Daten der Messlokation“","ZG6
  „Erwartete Daten der Messlokation“","ZG7
  „Im System vorhandene Daten der Messlokation“"
  "SG8 SEQ DE1229
  Erforderliches Messprodukt der Messlokation","Z19
  „Erforderliches Produkt der Messlokation“",--,"ZF4
  „Informative Erforderliches Produkt der Messlokation“",--,--
  "SG8 SEQ DE1229
  Zähleinrichtungsdaten",--,"Z03
  „Zähleinrichtungsdaten“","ZF5
  „Informative Zähleinrichtungsdaten“","ZA3
  „Erwartete Zähleinrichtungsdaten“","ZA4
  „Im System vorhandene Zähleinrichtungsdaten“"
  "SG8 SEQ DE1229
  OBIS-Daten der Zähleinrichtung / Smartmeter-Gateway",--,"Z20
  „OBIS-Daten der Zähleinrichtung“","ZF6
  „Informative OBIS-Daten der Zähleinrichtung“","ZA5
  „Erwartete OBIS-Daten der Zähleinrichtung“","ZA6
  „Im System vorhandene OBIS-Daten der Zähleinrichtung“"
  "SG8 SEQ DE1229
  Wandlerdaten",--,"Z04
  „Wandlerdaten“","ZF7
  „Informative Wandlerdaten“","ZB9
  „Erwartete Wandlerdaten“","ZC0
  „Im System vorhandene Wandlerdaten“"
  "SG8 SEQ DE1229
  Kommunikationseinrichtungsdaten",--,"Z05
  „Kommunikationseinrichtungsdaten“","ZF8
  „Informative Kommunikationseinrichtungsdaten“","ZB7
  „Erwartete Kommunikationseinrichtungsdaten“","ZB8
  „Im System vorhandene Kommunikationseinrichtungsdaten“"
  "SG8 SEQ DE1229
  Daten der technischen Steuereinrichtung",--,"Z06
  „Daten der technischen Steuereinrichtung“","ZF9
  „Informative Daten der technischen Steuereinrichtung“","ZC1
  „Erwartete Daten der technischen Steuereinrichtung“","ZC2
  „Im System vorhandene Daten der technischen Steuereinrichtung“"
  "SG8 SEQ DE1229
  Smartmeter-Gateway",--,"Z13
  „Smartmeter-Gateway“","ZG0
  „Informative Smartmeter-Gateway“","ZC3
  „Erwartetes Smartmeter-Gateway“","ZC4
  „Im System vorhandenes Smartmeter-Gateway“"
  "SG8 SEQ DE1229
  Daten der Steuerbox",--,"Z14
  „Steuerbox“",--,"ZH3
  „Erwartete Daten der Steuerbox“","ZH4
  „Im System vorhandene Daten der Steuerbox“"
  "SG8 SEQ DE1229
  Profildaten",--,"Z21
  „Profildaten“","ZG1
  „Informative Profildaten“","Z85
  „Erwartete Profildaten“","Z86
  „Im System vorhandene Profildaten“"
  "SG8 SEQ DE1229
  Profildaten der beteiligten Marktrolle",--,--,--,--,"Z33
  „Profildaten der beteiligten Marktrolle“"
  "SG8 SEQ DE1229
  Profilschardaten",--,"Z08
  „Profilschardaten“","ZG2
  „Informative Profilschardaten“","Z87
  „Erwartete Profilschardaten“","Z88
  „Im System vorhandene Profilschardaten“"
  "SG8 SEQ DE1229
  Referenzprofildaten",--,"Z38
  „Referenzprofildaten“","ZG3
  „Informative Referenzprofildaten“","Z89
  „Erwartete Referenzprofildaten“","Z90
  „Im System vorhandene Referenzprofildaten“"
  "SG8 SEQ DE1229
  Daten der Summenzeitreihe",--,"Z22
  „Daten der Summenzeitreihe“",--,--,--
  "SG8 SEQ DE1229
  Produkt-Daten der Summenzeitreihe",--,"Z23
  „Produkt-Daten der Summenzeitreihe“",--,--,--
  "SG8 SEQ DE1229
  „Daten der Überführungszeitreihe“",--,"Z24
  „Daten der Überführungszeitreihe“",--,--,--
  "SG8 SEQ DE1229
  Produkt-Daten der Überführungszeitreihe",--,"Z25
  „Produkt-Daten der Überführungszeitreihe“",--,--,--
  "SG8 SEQ DE1229
  Datenstand des ÜNB",--,--,--,--,"Z47
  „Datenstand des ÜNB“"
  "SG8 SEQ DE1229
  Datenstand des NB",--,--,--,--,"Z72
  „Datenstand des NB“"
  "SG8 SEQ DE1229
  Abgerechnete Daten der Bilanzkreissummenzeitreihe",--,"Z48
  „Abgerechnete Daten der Bilanzkreissummenzeitreihe des ÜNB“",--,--,--
  "SG8 SEQ DE1229
  Abgerechnete Daten der Bilanzierungsgebietssummenzeitreihe",--,"Z49
  „Abgerechnete Daten der Bilanzierungsgebietssummenzeitreihe des ÜNB“",--,--,--
  "SG8 SEQ DE1229
  Daten des Kunden des Lieferanten",--,"Z75
  „Daten des Kunden des Lieferanten“",--,"Z92
  „Erwartete Daten des Kunden des Lieferanten“","Z93
  „Im System vorhandene Daten des Kunden des Lieferanten“"
  "SG12 NAD DE3035
  Kunde des Lieferanten",--,"Z09
  „Kunde des LF“","Z65
  „Informativer Kunden des LF“","Z47
  „Erwarteter Kunde des LF“","Z48
  „Im System vorhandener Kunde des LF“"
  "SG12 NAD DE3035
  Korrespondenzanschrift des Kunden des Lieferanten",--,"Z04
  „Korrespondenzanschrift des Kunden des LF“","Z66
  „Informative Korrespondenzanschrift des Kunden des LF“","Z49
  „Erwartete Korrespondenzanschrift des Kunden des LF“","Z50
  „Im System vorhandene Korrespondenzanschrift des Kunden des LF“"
  "SG12 NAD DE3035
  Kunde des Messstellenbetreibers",--,"Z07
  „Kunde des MSB“",--,"Z39
  „Erwarteter Kunde des MSB“","Z40
  „Im System vorhandener Kunde des MSB“"
  "SG12 NAD DE3035
  Korrespondenzanschrift des Kunden des Messstellenbetreibers",--,"Z08
  „Korrespondenzanschrift des Kunden des MSB“",--,"Z41
  „Erwartete Korrespondenzanschrift des Kunden des MSB“","Z42
  „Im System vorhandene Korrespondenzanschrift des Kunden des MSB“"
  "SG12 NAD DE3035
  Kunde des Netzbetreibers",--,"Z25
  „Kunde des NB“","Z67
  „Informativer Kundenname des NB“","Z51
  „Erwarteter Kunde des Netzbetreibers“","Z52
  „Im System vorhandener Kunde des Netzbetreibers“"
  "SG12 NAD DE3035
  Korrespondenzanschrift des Kunden des Netzbetreibers",--,"Z26
  „Korrespondenzanschrift des Kunden des NB“","Z68
  „Informative Korrespondenzanschrift des Kunde des NB“","Z53
  „Erwartete Korrespondenzanschrift des Kunden des Netzbetreibers“","Z54
  „Im System vorhandene Korrespondenzanschrift des Kunden des Netzbetreibers“"
  "SG12 NAD DE3035
  Anschlussnehmer",--,"EO
  „Anschlussnehmer“","Z69
  „Informative Daten des Anschlussnehmers“","Z55
  „Erwarteter Anschlussnehmer“","Z56
  „Im System vorhandener Anschlussnehmer“"
  "SG12 NAD DE3035
  Beteiligter Marktpartner MP-ID",--,--,"VY
  „andere zugehörige Partei“",--,--
  "SG12 NAD DE3035
  Hausverwalter",--,"DDO
  „Hausverwalter“","Z70
  „Informative Daten des Hausverwalters“","Z57
  „Erwarteter Hausverwalter“","Z58
  „Im System vorhandener Hausverwalter“"
  "SG12 NAD DE3035
  Marktlokationsanschrift",--,"DP
  „Lieferanschrift“","Z63
  „Informative Marktlokationsanschrift“","Z59
  „Erwartete Marktlokationsanschrift“","Z60
  „Im System vorhandene Marktlokationsanschrift“"
  "SG12 NAD DE3035
  Messlokationsadresse",--,"Z03
  „Messlokationsadresse“","Z64
  „Informative Messlokationsadresse“","Z43
  „Erwartete Messlokationsadresse“","Z44
  „Im System vorhandene Messlokationsadresse“"
  "SG12 NAD DE3035
  Name und Adresse für die Ablesekarte",--,"Z05
  „Name und Adresse für die Ablesekarte“",--,"Z45
  „Erwarteter Name und Adresse für die Ablesekarte“","Z46
  „Im System vorhandener Name und Adresse für die Ablesekarte“"
  
  '''
# ---
